### PR TITLE
Represent closed in either end as closed in ADT 

### LINKED ISSUE #1563: MQTT streaming: internal Akka streams error triggers in test case prevents shutdown
### Expected Behavior

MQTT streams should shut down properly in test cases asserted by `assertAllStagesStopped` if the streams are completed and the MQTT session actors are shut down.

### Actual Behavior

Something triggers an error in the Akka Streams machinery which is evident in the logs and hinders proper shutdown.

### Relevant logs

```
2019-03-06 11:28:23,035 ERROR [mqtt-spec-akka.actor.default-dispatcher-3] [akka.actor.OneForOneStrategy]  56 (of class java.lang.Integer)
 scala.MatchError: 56 (of class java.lang.Integer)
     at akka.stream.impl.fusing.GraphInterpreter.$anonfun$toSnapshot$4(GraphInterpreter.scala:662)
```

### Reproducible Test Case

See `MqttSessionSpec` "MQTT server connector" should "close when no ping request received".

### LINKED ISSUE #26493: Akka Stream - StreamTestKit.printDebugDump - MatchError 50
I am using `StreamTestKit.printDebugDump(materializer.supervisor)` to check the topology of my stream graph. This gives me

```
scala.MatchError: 50 (of class java.lang.Integer)
	at akka.stream.impl.fusing.GraphInterpreter.$anonfun$toSnapshot$4(GraphInterpreter.scala:662)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	at scala.collection.TraversableLike.map(TraversableLike.scala:237)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:230)
	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:198)
	at akka.stream.impl.fusing.GraphInterpreter.toSnapshot(GraphInterpreter.scala:654)
	at akka.stream.impl.fusing.GraphInterpreterShell.toSnapshot(ActorGraphInterpreter.scala:648)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.$anonfun$applyOrElse$1(ActorGraphInterpreter.scala:755)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
	at scala.collection.immutable.Set$Set1.foreach(Set.scala:97)
	at scala.collection.TraversableLike.map(TraversableLike.scala:237)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:230)
	at scala.collection.AbstractSet.scala$collection$SetLike$$super$map(Set.scala:51)
	at scala.collection.SetLike.map(SetLike.scala:104)
	at scala.collection.SetLike.map$(SetLike.scala:104)
	at scala.collection.AbstractSet.map(Set.scala:51)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:755)
	at akka.actor.Actor.aroundReceive(Actor.scala:517)
	at akka.actor.Actor.aroundReceive$(Actor.scala:515)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:655)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:592)
	at akka.actor.ActorCell.invoke(ActorCell.scala:561)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)
	at akka.dispatch.Mailbox.run(Mailbox.scala:225)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
`50 == InClosed | OutClosed | Pulling` which is not covered by the matching function. Akka Stream is `"2.5.21"`, Scala 2.12 JVM.

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #26493

Should also fix https://github.com/akka/alpakka/issues/1563

### CODE REVIEW

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala [hunk 1]
```diff
@@ -668,10 +668,15 @@ import akka.stream.snapshot._
         logicSnapshots(logicIndexes(connection.inOwner)),
         logicSnapshots(logicIndexes(connection.outOwner)),
         connection.portState match {
-          case InReady  => ConnectionSnapshot.ShouldPull
-          case OutReady => ConnectionSnapshot.ShouldPush
-          case x if (x | InClosed | OutClosed) == (InClosed | OutClosed) =>
+          case InReady                                                     => ConnectionSnapshot.ShouldPull
+          case OutReady                                                    => ConnectionSnapshot.ShouldPush
+          case x if (x & (InClosed | OutClosed)) == (InClosed | OutClosed) =>
```
Comment (raboof) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: I'm not sure this logic is correct: if `x` is `InFailed | OutClosed`, `x & (InClosed | OutClosed)` would be `OutClosed` but `OutClosed` != `(InClosed | OutClosed)`...

Shouldn't the conditional be something like `if (x & (InClosed | OutClosed)) > 0` to match the comment?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala [hunk 2]
```diff
@@ -668,10 +668,15 @@ import akka.stream.snapshot._
...
+            // At least one side of the connection is closed: we show it as closed
             ConnectionSnapshot.Closed
+          case _ =>
+            // This should not be possible: connection alive and both push and pull enqueued but not received
```
Comment (raboof) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: Hard to say if this is correct but at least it's more explicit than before :)

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala (hunk 1 above)
Comment (raboof) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: Though the original code looks even worse: AFAICS `if (x | InClosed | OutClosed) == (InClosed | OutClosed)` means 'neither side is closed'...

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: From the docs up top
```
InFailed:  Always set in conjunction with InClosed. Indicates that the close event is a failure
```

So I think this is safe

Comment (raboof) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: :+1: on a closer look I agree!

**Review by raboof — APPROVED** [review]

**Pull request merged**